portal/portal-service-impl/impl/src/java/org/sakaiproject/portal/service/BullhornServiceImpl.java

```diff
@@ -0,0 +1,562 @@
+                                    for (String  to : site.getUsersIsAllowed(AnnouncementService.SECURE_ANNC_READ)) {
+                                        if (!from.equals(to) && !securityService.isSuperUser(to)) {
+                                            doAcademicInsert(from, to, event, ref, title, siteId, e.getEventTime(), url);
+                                            countCache.remove(to);
+                                        }
+                                    }
+                                }
+                            } catch (IdUnusedException idue) {
+                                log.error("No site with id '" + siteId + "'", idue);
+                            } finally {
+                                lock(sa);
+                            }
+                        } else if (AssignmentConstants.EVENT_ADD_ASSIGNMENT.equals(event)) {
+                            String siteId = pathParts[3];
+                            String assignmentId = pathParts[pathParts.length - 1];
+                            try {
+                                Assignment assignment = assignmentService.getAssignment(assignmentId);
+                                Time openTime = assignment.getOpenTime();
+                                if (openTime == null || openTime.getTime() < (new Date().getTime())) {
+                                    Site site = siteService.getSite(siteId);
+                                    String title = assignment.getTitle();
+                                    // Get all the members of the site with read ability
+                                    for (String  to : site.getUsersIsAllowed(AssignmentService.SECURE_ACCESS_ASSIGNMENT)) {
+                                        if (!from.equals(to) && !securityService.isSuperUser(to)) {
+                                            switchUser(to);
+                                            String url = assignmentService.getDeepLink(siteId, assignmentId);
+                                            doAcademicInsert(from, to, event, ref, title, siteId, e.getEventTime(), url);
+                                            countCache.remove(to);
+                                        }
+                                    }
+                                }
+                            } catch (IdUnusedException idue) {
+                                log.error("Failed to find either the assignment or the site", idue);
+                            }
+                        } else if (AssignmentConstants.EVENT_GRADE_ASSIGNMENT_SUBMISSION.equals(event)) {
+                            String siteId = pathParts[3];
+                            String submissionId = pathParts[pathParts.length - 1];
+                            try {
+                                SecurityAdvisor sa = unlock();
+                                AssignmentSubmission submission = assignmentService.getSubmission(submissionId);
+                                lock(sa);
```

Comment:
This should always be in a `finally` block so there's no way you can't remove the advisor.